openassessment/xblock/submission_mixin.py

```diff
@@ -200,6 +189,56 @@ def save_submission(self, data, suffix=''):  # pylint: disable=unused-argument
         else:
             return {'success': False, 'msg': self._(u"This response was not submitted.")}
 
+    def get_saved_file_metadata(self):
+        """
+        Get list of (file_download_url, file_description, file_name, file_size) for all uploaded files.
+        Deleted files are represented as (None, None, None, 0)
+        """
+        try:
+            saved_files_descriptions = json.loads(self.saved_files_descriptions)
+            saved_files_names = json.loads(self.saved_files_names)
+            saved_files_sizes = json.loads(self.saved_files_sizes)
+        except ValueError:
+            saved_files_descriptions = []
+            saved_files_names = []
+            saved_files_sizes = []
+
+        file_metadata = []
+
+        if saved_files_descriptions == []:
+            # This is the old behavior, required for a corner case and should be eventually removed.
+            # https://github.com/edx/edx-ora2/pull/1275 closed a loophole that allowed files
+            # to be uploaded without descriptions. In that case, saved_files_descriptions would be
+            # an empty list. If there are currently users in that state who have files uploaded
+            # with no descriptions but have not yet submitted, they will fall here.
+            for i in range(self.MAX_FILES_COUNT):
+                file_url = self._get_download_url(i)
+                file_description = ''
+                file_name = ''
+                file_size = 0
+                if file_url:
+                    try:
+                        file_description = saved_files_descriptions[i]
+                        file_name = saved_files_names[i]
+                        file_size = saved_files_sizes[i]
+                    except IndexError:
+                        pass
+                    file_metadata.append((file_url, file_description, file_name, file_size))
+                else:
+                    break
+        else:
+            zipped_metadata = zip(saved_files_descriptions, saved_files_names, saved_files_sizes)
+            for i, (file_description, file_name, file_size) in enumerate(zipped_metadata):
+                if file_description is None:
+                    # We are passing Nones to the template because when files are deleted, we still want to
```

Comment:
In our future world of being able to track indices of file objects more easily, we could just pass an index to the front-end to let it know how to correctly create the download/upload URLs.